Preparatory work for eResistance fix

OnlineCop opened the pull request:
# I Tried Really, Really Hard...

Yay. One more _massive_ update that changes all the things, all over the place.

## Organization
I tried my best to break the commits down into individual "change only related things" chunks, like renaming a variable, or adding comments for clarification.

## Common Names
Spells (`s_spell`), items (`s_item`), and fighters (`KFighter`) (usually used in the context of `party[...]`) each have items that can cast (or cause) elemental effects on the target. I renamed these to `spell_elemental_effect`, `item_elemental_effect`, and `weapon_elemental_effect`, respectively.
- These were the ones that were "sometime" 0-based and other times 1-based. I might have even missed a few conversions due to variables getting passed around all over the place.

## Now For the Oh-No!

Something to note with this commit: the arrays defined in **res.cpp** have not been changed in this commit. That's an upcoming PR.

### This means that the data values may be all wonky!

However, the next PR, which actually changes the values in **res.cpp** is following closely on the heels of this one.
- I'm trying to go through **every** ***single*** **item** to ensure that the values make sense.
- Having enum names in place of Magic Numbers should speed that along.
- I'm also not averse to committing that PR and letting "the community" give all the values a good going-over. That should help spot and fix inaccuracies.

Since the player (and therefore, the save games) are probably using values that were originally also a blend of 0-based and 1-based for a lot of these stats, *you may need to start a fresh game* in order to do thorough testing, before reporting bugs.

OnlineCop committed:
utilities.h: Change snprintf() return type from void to int

OnlineCop committed:
Combine resabil.mon, allstat.mon

Appends resabil.mon onto the end of the allstat.mon lines.
- Removes both unused Index columns.
- Excludes second Name column (from resabil.mon).
Moves all KFighter read/write out of KEnemy::LoadEnemies() and into KFighter iostream operators.

OnlineCop committed:
utilities.h: add read() and write() overloads for iostream

Problem: int8_t and uint8_t were being parsed by operator>> and operator<< as though they were chars instead of numerics.
Fix: cast int8_t and uint8_t types to int, so numbers in text get converted to their numerical values instead of their ASCII values.

OnlineCop committed:
kq.h: Remove unused savedir

OnlineCop committed:
EMagic: Add M_NONE=0 and M_TOTAL entries to enum

OnlineCop committed:
Rename EMagic enum to eMagic (consistent with other enums)

OnlineCop committed:
s_item::name, s_item::desc: to std::string

OnlineCop committed:
Rename s_item::name to s_item::item_name

OnlineCop committed:
Rename s_item::desc to s_item::item_desc

OnlineCop committed:
Rename s_item::elem to s_item::item_elemental_effect

OnlineCop committed:
Explicitly set KFighter::bstat to an eStat value

Because who doesn't love vague magic numbers?

OnlineCop committed:
Add eWeaponRestrict enum

s_item::hnds can be either 0 or 1, for weapon/shield/armor/etc. types,
or it can be en eMagic enum value for spell book types.

Not confusing at all, huh?

OnlineCop committed:
Rename s_item::bst to s_item::seed_stat

OnlineCop committed:
Update s_item comments

OnlineCop committed:
eWeapon: Add in missing icon range

OnlineCop committed:
Rename s_spell::elem to s_spell::spell_elemental_effect

OnlineCop committed:
KPlayer: spells[] comment

OnlineCop committed:
KFighter: Rename welem to weapon_elemental_effect

OnlineCop committed:
Change all *_elemental_effect from 1-based to 0-based

This does NOT include the data!
That is, res.cpp is still using a mixture of 0- and 1-based values.

OnlineCop commented on include/utilities.h:
```diff
@@ -38,22 +38,32 @@
  * \param[out] dest String to write final output to.
  * \param[in] format Text containing "%..." style notation to expand.
  * \param[in] args... Zero or more arguments to pass into snprintf().
+ * \returns On success: number of bytes written, not including terminating null character;
```
This is essentially mirroring what sprintf() or snprintf() would return.

OnlineCop commented on src/enemyc.cpp:
```diff
@@ -642,176 +642,43 @@ void KEnemy::LoadEnemies(const std::string& fullPath, Raster* enemy_gfx)
-        int imagefile_y_coord = 0;
-        iss >> imagefile_y_coord;
+        KFighter fighter;
+        iss >> fighter;
```
This moved all the `iss >> fighter.XYZ;` calls into the KFigher class, since it should be responsible for its own reading/writing rules.

OnlineCop commented on src/enemyc.cpp:
```diff
@@ -642,176 +642,43 @@ void KEnemy::LoadEnemies(const std::string& fullPath, Raster* enemy_gfx)
-        iss >> fighter_loaded_from_disk.imb[0];
-        iss >> fighter_loaded_from_disk.imb[1];
+        fighter.img = std::make_shared<Raster>(fighter.cw, fighter.cl);
+        enemy_gfx->blitTo(fighter.img.get(), fighter.cx, fighter.cy, 0, 0, fighter.cw, fighter.cl);
```
One of the *could-be-significant* changes made here, was that `KFighter::cx` and `KFighter::cy` were never set before, and the positions within the source image were read into temporary variables and only used in this `enemy_gfx->blitTo()` call.

Those `cx` and `cy` values are actually used throughout the game as temporaries, such as when the enemy's attack stance changes or animates.

So I figured that, as they're always being written to by those other places in code before they're being used, whatever values were here would just get clobbered anyway, so we might as well use them to temporarily store those values.

OnlineCop commented on src/fighter.cpp:
```diff
@@ -21,72 +21,53 @@
+    , aframe {}
+    , crit {}
+    , defend {}
+    , ai {}
+    , aip {}
+    , atrack {}
```
Turns out, default-initializing an array DOES zero the whole thing out.

pedro-w replied:
Hm i did not realise that!

OnlineCop commented on include/res.h:
```diff
@@ -45,14 +46,14 @@
-    /*! Name of the item (16 chars plus '\0') */
-    char name[17];
+    /*! Item's short name (around 16 characters). */
+    std::string name;
```
I looked as carefully as possible that there's no C memory management going on with `s_item` (like memcpy, memmove, strcpy*, strlen, etc.), but I may have overlooked a few.

If we suddenly get strange crashes due to strings, they almost never manifest at the place of the *actual* error; they tend to abort when something that seems unrelated goes out of scope and it causes a double-delete or something.

\* The C version of `strcpy()`, not the `std::string` overloaded one that I added in earlier.

pedro-w replied:
Got an error relating to this:
```
In file included from /Projects/kq-fork/include/kq.h:47:
/Projects/kq-fork/include/utilities.h:50:60: error: cannot pass object of non-trivial type 'std::__1::basic_string<char>' through variadic function; call will abort at runtime [-Wnon-pod-varargs]
    int size_s = std::snprintf(nullptr, 0, format.c_str(), args...);
                                                           ^
/Projects/kq-fork/src/combat.cpp:1322:21: note: in instantiation of function template specialization 'sprintf<std::__1::basic_string<char> >' requested here
                    sprintf(strbuf, _("%s found!"), items[found_item].item_name);
                    ^
```
I think possibly needs to be `items[found_item].item_name.c_str()` ?

pedro-w replied:
Also
* hskill.cpp:794
* intrface.cpp:1435
* intrface.cpp:1439
* intrface.cpp:1451
* intrface.cpp:1455

OnlineCop replied:
Looking into this now. Frustrated that my compiler gave zero warnings about this.

OnlineCop replied:
The latest push should take care of this.

And in doing so, I noticed that `s_item::item_name` was changed to a std::string, but `s_spell::name` was not. I'll probably want to do that...

pedro-w commented on src/menu.cpp:
```diff
@@ -742,9 +742,9 @@ KFighter player2fighter(int who)
      * it has no other power to begin with (the "weapon_elemental_effect" property
      * is 1-based: value of 0 means "no imbue".
```
This comment doesn't match the code any more

OnlineCop replied:
Fixed; thanks for spotting it!

pedro-w commented:
I may have got a bit mixed up but `R_TOTAL_RES` means "no elemental effect" doesn't it? In which case should we rename it to something more descriptive? I noticed that most enums have got a `NUM_XXX // always last` line,  apart from eResistance. How about 
```
    R_NONE = 16,

    NUM_RESISTANCES = R_NONE // always last
```
for consistency

OnlineCop committed:
utilities.h: Simplify sprintf()

OnlineCop committed:
Fix sprintf() for item_name (std::string)

OnlineCop committed:
menu.cpp: Update weapon_elemental_effect comment

- Also removes commented-out for() loop that had something to do with opal_power calculation.

OnlineCop commented:
> I may have got a bit mixed up but `R_TOTAL_RES` means "no elemental effect" doesn't it? In which case should we rename it to something more descriptive? I noticed that most enums have got a `NUM_XXX // always last` line, apart from eResistance. How about
> 
> ```
>     R_NONE = 16,
> 
>     NUM_RESISTANCES = R_NONE // always last
> ```
> 
> for consistency

I think this approach would be fine. We still need `NUM_RESISTANCES` since it's used as array size initialization, but would you (or anyone else) be averse to renaming all of the last elements in enums as `<PREFIX>_TOTAL`?

- NUM_RESISTANCES => R_TOTAL
- NUM_BUBBLE_STYLES => B_TOTAL
- NUM_BUBBLE_STEMS => STEM_TOTAL
- NUM_COMMANDS => COMMAND_TOTAL
- etc.

Basically, all the other enum values share a common prefix (with only 2-3 notable exceptions), so it seems like it would make more sense to have the same prefix for the "last"/"count" element in the enums.

pedro-w commented:
>  would you (or anyone else) be averse to renaming all of the last elements in enums as `<PREFIX>_TOTAL`?

I'm not really bothered but I think if it's `NUM_XXX` then it's clearly not one of the enum members. Say if I saw `B_TOTAL` I might think it's one of the bubble styles

OnlineCop commented:
To keep the number of commits/changes to this PR as minimal as possible, I'm fine adding the R_NONE to the enum and some of the equality checks, but I'm going to avoid all the other changes till after this is merged.

I hate submitting PRs with dozens of commits like this, since it's a nightmare for the 2-3 of us to wade through.

OnlineCop committed:
Add eResistance::R_NONE to enum
